osu.Game/Screens/SelectV2/Carousel.cs

```diff
@@ -0,0 +1,444 @@
+        /// </remarks>
+        public virtual object? CurrentSelection
+        {
+            get => currentSelection;
+            set
+            {
+                if (currentSelectionCarouselItem != null)
+                    currentSelectionCarouselItem.Selected.Value = false;
+
+                currentSelection = value;
+
+                currentSelectionCarouselItem = null;
+                currentSelectionYPosition = null;
+                updateSelection();
+            }
+        }
+
+        private List<CarouselItem>? displayedCarouselItems;
+
+        private readonly DoublePrecisionScroll scroll;
+
+        protected Carousel()
+        {
+            InternalChildren = new Drawable[]
+            {
+                new Box
+                {
+                    Colour = Color4.Black,
+                    RelativeSizeAxes = Axes.Both,
+                },
+                scroll = new DoublePrecisionScroll
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Masking = false,
+                }
+            };
+
+            Items.BindCollectionChanged((_, _) => QueueFilter());
+        }
+
+        /// <summary>
+        /// Queue an asynchronous filter operation.
+        /// </summary>
+        public void QueueFilter() => Scheduler.AddOnce(() => filterTask = performFilter());
+
+        /// <summary>
+        /// Create a drawable for the given carousel item so it can be displayed.
+        /// </summary>
+        /// <remarks>
+        /// For efficiency, it is recommended the drawables are retrieved from a <see cref="DrawablePool{T}"/>.
+        /// </remarks>
+        /// <param name="item">The item which should be represented by the returned drawable.</param>
+        /// <returns>The manifested drawable.</returns>
+        protected abstract Drawable GetDrawableForDisplay(CarouselItem item);
+
+        /// <summary>
+        /// Create an internal carousel representation for the provided model object.
+        /// </summary>
+        /// <param name="model">The model.</param>
+        /// <returns>A <see cref="CarouselItem"/> representing the model.</returns>
+        protected abstract CarouselItem CreateCarouselItemForModel(T model);
+
+        #region Filtering and display preparation
+
+        private Task filterTask = Task.CompletedTask;
+        private CancellationTokenSource cancellationSource = new CancellationTokenSource();
+
+        private async Task performFilter()
+        {
+            Debug.Assert(SynchronizationContext.Current != null);
+
+            var cts = new CancellationTokenSource();
+
+            lock (this)
+            {
+                cancellationSource.Cancel();
+                cancellationSource = cts;
+            }
+
+            Stopwatch stopwatch = Stopwatch.StartNew();
+            IEnumerable<CarouselItem> items = new List<CarouselItem>(Items.Select(CreateCarouselItemForModel));
```

Comment:
This sort of thing worries me. In memory profiling this is pretty huge, although it's probably not avoidable...

![1736764729](https://github.com/user-attachments/assets/492cc345-c5e3-4b6f-ad11-8c9862c33e3a)

However the more important and actionable part is that these are going to allocate on the update thread, are they not? This is before any `await`s, so this part of the method still executes synchronously. This should probably be on a background thread, at least after a certain size. Would also be nice to have cancellation support for this to possibly avoid trashing the heap too much.

Reply:
I guess we can move the transform (`Select`) part, but the copy needs to be locked or update thread. Just keeping simple for now.

Reply:
> but the copy needs to be locked or update thread

Well I guess this means this review thread just plain cannot be actioned upon because one way or another the update thread will have to bear the overhead of a copy of 200k+ objects in the worst cases...? In fact at this point you might as well not change anything because the `.Select()` here is lazy and thus is performing the functions of *both* a transform and a copy, so splitting it into two parts would only increase the overhead...

I dunno, I really think this *is* going to be a performance bottleneck we'll have to address at some point somehow.

At least maybe if the copy process could be aborted midway through it would be *some* solace...

Reply:
There's many ways we can improve of this, but it's already much faster than the old carousel since the majority is now async.

The copy overhead wasn't showing up in my profiling so I haven't addressed it yet.

Reply:
Well okay if you think it's not an issue I'm willing to go along with it for now.